internal/watcher/watcher_test.go

```diff
@@ -0,0 +1,446 @@
+
+	temp := t.TempDir()
+	destKeep := filepath.Join(temp, "keep")
+	destRemove := filepath.Join(temp, "remove")
+
+	testutils.Copy(t, filepath.Join("testdata", "withsubdir"), destKeep)
+	testutils.Copy(t, filepath.Join("testdata", "withsubdir"), destRemove)
+
+	// Instantiate the object
+	w, err := watcher.New(context.Background(), []string{destRemove, destKeep})
+	require.NoError(t, err, "Setup: Can't create watcher")
+
+	// Start it
+	err = w.Start(mockService{})
+	require.NoError(t, err, "Setup: Can't start watcher")
+	defer w.Stop(mockService{})
+
+	// Check GPT versions on the 2 dirs
+	assertGPTVersionEquals(t, destKeep, 2)
+	assertGPTVersionEquals(t, destRemove, 2)
+
+	// Give some unexisting directories to watch
+	err = w.UpdateDirs(context.Background(), []string{destKeep, "unexisting"})
+	require.Error(t, err, "UpdateDirs should have failed but didn't")
+
+	// Modify files in previous watched directories
+	writeToFiles(t, []string{
+		filepath.Join(destKeep, "alreadyexists"),
+		filepath.Join(destRemove, "alreadyexists")})
+
+	// Stop the watcher
+	err = w.Stop(mockService{})
+	require.NoError(t, err, "Can't stop watcher")
+
+	// compare GPT.ini version: only keep and add should be updated
+	testutils.WaitForWrites(t)
+	assertGPTVersionEquals(t, destKeep, 3)
+	assertGPTVersionEquals(t, destRemove, 3)
+}
+
+func TestStopWithoutStart(t *testing.T) {
+	t.Parallel()
+
+	temp := t.TempDir()
+	w, err := watcher.New(context.Background(), []string{temp})
+	require.NoError(t, err, "Can't create watcher")
+
+	var s service.Service = mockService{}
+
+	err = w.Stop(s)
+	require.Error(t, err, "Stop should have failed but hasn't")
+}
+
+func assertGPTVersionEquals(t *testing.T, path string, version int) {
+	t.Helper()
+
+	testutils.WaitForWrites(t)
+	gptfile := filepath.Join(path, "gpt.ini")
+
+	// Wait for GPT.ini to be created
+	var gptFileExists bool
+	if _, err := os.Stat(gptfile); err == nil {
+		gptFileExists = true
+	}
+
+	if version == 0 {
+		require.False(t, gptFileExists, "GPT.ini should not exist")
+		return
+	}
+	require.True(t, gptFileExists, "GPT.ini not created")
+
+	cfg, err := ini.Load(gptfile)
+	require.NoError(t, err, "Can't load GPT.ini")
+
+	v, err := cfg.Section("General").Key("Version").Int()
+	require.NoError(t, err, "Can't get GPT.ini version as an integer")
+
+	assert.Equal(t, version, v, "GPT.ini version is not equal to the expected one")
+}
+
+func assertGPTVersionError(t *testing.T, path string) {
+	t.Helper()
+
+	testutils.WaitForWrites(t)
+	gptfile := filepath.Join(path, "gpt.ini")
+
+	// Wait for GPT.ini to be created
+	var gptFileExists bool
+	if _, err := os.Stat(gptfile); err == nil {
+		gptFileExists = true
+	}
+	require.True(t, gptFileExists, "GPT.ini not created")
+
+	cfg, err := ini.Load(gptfile)
+	require.NoError(t, err, "Can't load GPT.ini")
+
+	_, err = cfg.Section("General").Key("Version").Int()
+	require.Error(t, err, "Version should be invalid")
+}
+
+func writeToFiles(t *testing.T, files []string) {
```

Comment:
maybe a better name would `updateFiles()`

(Almost tempted to name it `updateFilesWithNEWCONTENT` :p)